node/lifecycle_test.go

```diff
@@ -0,0 +1,211 @@
+
+	eb := record.NewBroadcaster()
+	eb.StartLogging(log.G(ctx).Infof)
+	eb.StartRecordingToSink(&corev1client.EventSinkImpl{Interface: client.CoreV1().Events(testNamespace)})
+	fakeRecorder := &fakeDiscardingRecorder{}
+
+	secretInformer := scmInformerFactory.Core().V1().Secrets()
+	configMapInformer := scmInformerFactory.Core().V1().ConfigMaps()
+	serviceInformer := scmInformerFactory.Core().V1().Services()
+
+	mockProvider, err := mock.NewMockProviderMockConfig(mock.MockConfig{}, testNodeName, "linux", "1.2.3.4", 0)
+	assert.NilError(t, err)
+	config := PodControllerConfig{
+		PodClient:         client.CoreV1(),
+		PodInformer:       podInformer,
+		EventRecorder:     fakeRecorder,
+		Provider:          mockProvider,
+		ConfigMapInformer: configMapInformer,
+		SecretInformer:    secretInformer,
+		ServiceInformer:   serviceInformer,
+	}
+
+	pc, err := NewPodController(config)
+	assert.NilError(t, err)
+
+	p := corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "my-pod",
+			Namespace:       testNamespace,
+			UID:             "4f20ff31-7775-11e9-893d-000c29a24b34",
+			ResourceVersion: "100",
+		},
+		Spec: corev1.PodSpec{
+			NodeName: testNodeName,
+		},
+		Status: corev1.PodStatus{
+			Phase: corev1.PodPending,
+		},
+	}
+
+	listOptions := metav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector("metadata.name", p.ObjectMeta.Name).String(),
+	}
+
+	watchErrCh := make(chan error)
+
+	watcher, err := client.CoreV1().Pods(testNamespace).Watch(listOptions)
+	assert.NilError(t, err)
+	go func() {
+		_, watchErr := watchutils.UntilWithoutRetry(ctx, watcher,
+			// Wait for the pod to be created
+			// TODO(Sargun): Make this "smarter" about the status the pod is in.
+			func(ev watch.Event) (bool, error) {
+				pod := ev.Object.(*corev1.Pod)
+				return pod.Name == p.ObjectMeta.Name, nil
+			})
+
+		watchErrCh <- watchErr
+	}()
+
+	go podInformerFactory.Start(ctx.Done())
+	go scmInformerFactory.Start(ctx.Done())
+
+	_, e := client.CoreV1().Pods(testNamespace).Create(&p)
+	assert.NilError(t, e)
+
+	assert.NilError(t, <-watchErrCh)
+
+	podControllerErrCh := make(chan error, 1)
+	go func() {
+		podControllerErrCh <- pc.Run(ctx, podSyncWorkers)
+	}()
+
+	watcher, err = client.CoreV1().Pods(testNamespace).Watch(listOptions)
+	assert.NilError(t, err)
+	go func() {
+		_, watchErr := watchutils.UntilWithoutRetry(ctx, watcher,
+			// Wait for the pod to be started
+			func(ev watch.Event) (bool, error) {
+				pod := ev.Object.(*corev1.Pod)
+				return pod.Status.Phase == corev1.PodRunning, nil
+			})
+
+		watchErrCh <- watchErr
+	}()
+
+	select {
+	case <-ctx.Done():
+		t.Fatal("Context ended early")
+	case err = <-podControllerErrCh:
+		assert.NilError(t, err)
+	case err = <-watchErrCh:
+		assert.NilError(t, err)
+
+	}
+
+	watcher, err = client.CoreV1().Pods(testNamespace).Watch(listOptions)
```

Comment:
Is there a race here with the watcher?  Can we miss the deleted event or do events wait for watchers?

Reply:
Events are delivered to watchers. If no watcher or watcher loses the event, a lister is to be used to fetch latest state. Therefore, events do not demand delivery guarantees and controllers should account for that.